README.md
# mooink

client/changelog.txt
7/29/2021:
* Added tooltips to join screen
* Modified color selector (put 3 or more points into vanity to pick color)
* Re-wrote join screen, now with twice as much flexbox
    * Imagine actually learning CSS lmao

7/29/2021:
* Optimized code for better GC performance, for server & client
* Modified shadows code to look smoother
    * Players used to only appear if more than half their character was
        outside a shadow. Now, they appear whenever a tiny part of their 
        body is outside a shadow (and the part behind the shadow gets 
        covered up client-side, instead of enforced server-side)

7/26/2021:
* Fixed shadows dissapearing bug
* Bots now cannot effectively double-team you until you have 3 or more kills
    * Zombies are an exception, but you control the spawning of zombies so it doesn't matter that much
    * Gun bots might appear to be double-teaming, but only one of them is actually shooting at a time,
        i.e. they cannot **effectively** double-team you
* Gun bots now follow you and will automatically place a spike if trapped
* Added zombies (very high resistance NPC, but cannot heal. Follows you constantly and attacks with melee)
* Added new spawner structure
    > Spawns zombies that attack you
* Fixed issue of healthbar being covered by other player

7/24/2021:
* Fixed gameloop (gameplay should be smoother)
    * Further optimizations are still pending
* Added damage texts
* Texture load now starts before the page is fully loaded, so everything will load faster
* Removed shadows caching
    * Fixes shadows bug
    > With new optimizations to the occludes function, shadow calculations
        are extremely fast, so caching really just causes gc issues
* Added webhook for telemetry

7/23/2021:
* Collisions now look a lot cleaner than before
    * Switched from extrapolation to interpolation
* Switched to a custom gameloop timer for more accuracy
    * Old setInterval timer was consistently ~15 ms too slow
* Fixed lag spike caused by improper setIntermediate 

7/22/2021:
* Fixed bug where boxes containing different items would merge on drag
* Inventory now shows images instead of text (more compact + prettier)
* Optimizations to game join menu and inventory image loading (caching)
* Added recipes menu
* Continued re-texturing work

7/20/2021:
* Nerfed nuro heavily
    * Nuro starts off as being unable to kill a stationary player as long as you heal, and difficulty
        scales with the number of kills you have. At 10 kills, nuro's difficulty is maxed out and is
        just as hard as the original nuros
* Fixed the "player flickering as it comes into view" bug
* Removed serialized player caching, it literally only saves a few microseconds, but causes huge GC issues
    > don't ask why I added it in the first place, let's just say I messed up a few unit conversions

7/19/2021:
* Made game menu look better
* Nerfed wrench speed by a little
* "Breaching rifle" is now replaced by "rifle"
    * Deals 0 structures damage, but 60 player damage
* Switched to ws, which should be (slightly) faster
* Fixed structure persistance issue

7/17/2021: 
* Fixed the "undefined leaderboard" glitch
* You no longer gain points for kills made while you're dead
* Respawning no longer provides you with an extra crafter

Entry 9.5 (patch): Improved box texture (now you can see what's in a box after it's placed)

Entry 9.5: 
* Retextured structures
* Hit animations now look prettier
* Various optimizations
* Fixed a few box exploits (e.g. "get boost wrench, box your boost pads and give to a trap player for infinite traps")
* Added new structure: Lapis spikes, which deals damage when you hit them. 
    Only available through crafting

Entry 9:
* Fixed lag caused by bots (obviously, doesn't apply to bot-free version)
* Made shadows look better (shadows are now on the top layer)
* Fixed bug where inventory doesn't display after respawn

Entry 8.5:
* Modified position backtracking
    * Improves trap mechanics a bit
* Extended position backtracking to all objects, so collisions are more accurate
haha low physics tickrate go brr
* You can now craft a new crafting table

Entry 8:
* Crafting added
    * Place items onto the crafting bench, then hit once with any weapon to craft
    * Craft items will get added to EITHER inventory or actionbar, so make sure to check both after crafting
    * Check the crafting page (link can be found on bottom right of game screen) for recipies and more info
    * Crafting right now is limited in the items that can be made, but it will be greatly expanded in the future
* Build placements when moving at high speed (i.e. boosting) are now more accurate

Entry 7.5: respawn inventory bug fixed

Entry 7 (7/11/2021):
* UI overhaul - text is FAR more readable, and all the menus are more obvious
* Inventory system is added! Click "inventory" to open the menu, and drag to 
    re-order inventory items
Crafting update will come soon

Entry 6:
* Trench is no longer a secondary - Estick is the new one, granting access to Trench
    * This is done in preparation for crafting update
* Hammer is now a secondary, with the same damage stats as wrench (but same speed as stick)
    * More moomoo-like feel

Entry 5 (7/8/2021):
* Tweaked trap mechanics, now the capture is more "smooth" (no speed increase)
    * Unfortunately, this adds a "rubber-banding" like effect, which isn't present
        in moomoo, but the goal here is to eliminate the "stuck in center" 
        case
            * Moomoo benefits from a higher physics tick rate, due to better
                servers, so this is unavoidable for a repl-hosted game. Time
                rewinding helps this a bit, however
* Game will rewind time for trap collisions to get more accurate measurements
    * Only will make a difference for high speed (e.g. wrench w/ max speed, or boost pad)
    * Physics are now consistent for all movement speeds

Entry 4 (7/7/2021):
* Added a bot-free server, available at nobot.mooink.repl.co

common/msgpack.js
(function(exports) {
    var the_encoder = new TextEncoder();
    var the_decoder = new TextDecoder();

    const TWONUM_IDENT = 2;
    const SMALLNUM_BASE = 40;
    const SMALLNUM_LENGTH = 200;

    const ARRAY_IDENT = 1;
    const SMALLARRAY_BASE = SMALLNUM_BASE + SMALLNUM_LENGTH;
    const SMALLARRAY_LENGTH = 10;

    const STRING_IDENT = 3;
    const FLOAT_IDENT = 4;
    const NEG_FLOAT_IDENT = 5;
    const FLOAT_PRECISION = 1000; // preserve 3 digits

    const NEW_CLAN_DELIM = 18;
    const CLAN_DESTROY_DELIM = 22;

    const PLAYER_UPDATE_DELIM = 19;
    const STRUCTURE_DELIM = 20;
    const PLAYER_DESTROY_DELIM = 23;
    const STRUCTURE_DESTROY_DELIM = 24;
    const PLAYER_NAME_UPDATE = 25;
    const PING = 26;
    const CHAT = 27;
    const TOGGLEAUTOHIT = 28;
    const BULLET_DELIM = 29;
    const BULLET_DESTROY_DELIM = 30;

    const CLAN_REQUEST_DELIM = 31;
    exports.SCORE_UPDATE_DELIM = 32;
    exports.LEADERBOARD_UPDATE_DELIM = 33;
    
    const HIT_DELIM = 34;
    exports.HIT_DELIM = HIT_DELIM;

    exports.TEAM_LOCATION_DELIM = 35;
    const MAP_PING = 36;
    exports.MAP_PING = MAP_PING;
    
    exports.RECONNECT = 37;
	exports.MISC = 38;

    const CLICK = 6;
    const ROTATE = 7;
    const SELITEM = 8;
    const LEFT = 9;
    const RIGHT = 10;
    const UP = 11;
    const DOWN = 12;
    const UPPERLEFT = 13;
    const UPPERRIGHT = 14;
    const LOWERLEFT = 15;
    const LOWERRIGHT = 16;
    const MOVERELEASE = 17;
    const JOIN = 0;

    const SMALLNUM_MIDDLE = Math.round(SMALLNUM_LENGTH / 2);

    function addNumb(acc, num) {
        if (num >= 0) {
            if ((num >= 0) && (num < SMALLNUM_LENGTH)) {
                acc.push(num + SMALLNUM_BASE);
            }
            else if ((num >= 0) && (num <= (256 * 255 + 255))) {
                acc.push(TWONUM_IDENT);
                let rem = num % 256;
                acc.push((num - rem) / 256);
                acc.push(rem);
            }
        }
        return acc;
    }

    function addHit(acc) {
        acc.push(HIT_DELIM);
    }

    function addPing(acc) {
        acc.push(MAP_PING);
    }

    function addPosNegNumb(acc, num) {
        return addNumb(acc, Math.max(0, Math.min(SMALLNUM_LENGTH, num + SMALLNUM_MIDDLE)));
    }

    function addString(acc, input) {
        if (input.length <= 255) {
            acc.push(STRING_IDENT);
            let encoded = the_encoder.encode(input);
            acc.push(encoded.length);
            acc.push(...encoded);
        }
        return acc;
    }

    function addFloat(acc, num) {
        if (num < 0) {
            acc.push(NEG_FLOAT_IDENT);
            let actualNum = Math.round(Math.abs(num) * FLOAT_PRECISION);
            let rem = actualNum % 256;
            acc.push((Math.abs(actualNum) - rem) / 256);
            acc.push(rem);
            return acc;
        }
        else {
            acc.push(FLOAT_IDENT);
            let actualNum = Math.round(num * FLOAT_PRECISION);
            let rem = actualNum % 256;
            acc.push((actualNum - rem) / 256);
            acc.push(rem);
            return acc;
        }
    }

    function addArray(acc, length) {
        if (length <= SMALLARRAY_LENGTH) {
            acc.push(SMALLARRAY_BASE + length);
        }
        else {
            acc.push(ARRAY_IDENT);
            acc.push(length);
        }
        return acc;
    }

    function expectNumb(input) {
        let data = input.data;
        if (data.length - input.start >= 1) {
            if ((data[input.start] >= SMALLNUM_BASE) && (data[input.start] < (SMALLNUM_BASE + SMALLNUM_LENGTH))) {
                input.start += 1;
                return data[input.start-1] - SMALLNUM_BASE;
            }
            else if (data[input.start] == TWONUM_IDENT) {
                if (data.length - input.start >= 3) {
                    input.start += 3;
                    return 256 * data[input.start-2] + data[input.start - 1];
                }
            }
            else if (data[input.start] == FLOAT_IDENT) {
                if (data.length - input.start >= 3) {
                    input.start += 3;
                    return (256 * data[input.start-2] + data[input.start - 1]) / 1000.0;
                }
            }
            else if (data[input.start] == NEG_FLOAT_IDENT) {
                if (data.length - input.start >= 3) {
                    input.start += 3;
                    return (-1.00) * (256 * data[input.start-2] + data[input.start - 1]) / 1000.0;
                }
            }
        }
        return undefined;
    }

    function expectHit(input) {
        let data = input.data;
        if (data.length - input.start >= 1) {
            if (data[input.start] == HIT_DELIM) {
                input.start += 1;
                return true;
            }
        }
        return false;
    }

    function expectPing(input) {
        let data = input.data;
        if (data.length - input.start >= 1) {
            if (data[input.start] == MAP_PING) {
                input.start += 1;
                return true;
            }
        }
        return false;
    }

    function expectString(input) {
        let data = input.data;
        if (data.length - input.start >= 2) {
            if (data[input.start] == STRING_IDENT) {
                let length = data[input.start + 1];
                if (data.length - input.start >= (2 + length)) {
                    input.start += 2 + length;
                    return the_decoder.decode(data.slice(input.start - length, input.start));
                }
            }
        }
        return undefined;
    }

    function expectPosNegNumb(input) {
        let dat = expectNumb(input);
        if (typeof dat !== 'undefined') {
            return dat - SMALLNUM_MIDDLE;
        }
        return undefined;
    }

    // We don't worry too much about processArray leaving the message in a broken state if it fails because processArray is not supposed to fail in the first place
    /*
    function processArray(input, length, offset) {
        let toReturn = [];
        for (let j = 0; j < length; j++) {
            let temp = undefined;
            if ((temp = expectNumb(input)) != undefined) {
                toReturn.push(temp);
            }
            else if ((temp = expectString(input)) != undefined) {
                toReturn.push(temp);
            }
            else {
                return undefined;
            }
        }
        return toReturn;
    }
    */

    function expectArray(input) {
        let data = input.data;
        if (data.length > input.start) {
            if ((data[input.start] >= SMALLARRAY_BASE) && (data[input.start] < SMALLARRAY_BASE + SMALLARRAY_LENGTH)) {
                let length = data[input.start] - SMALLARRAY_BASE;
                if (data.length - input.start >= 1) {
                    input.start += 1;
                    return length;
                }
            }
            else if (data[input.start] == ARRAY_IDENT) {
                if (data.length - input.start >= 2) {
                    let length = data[input.start + 1];
                    input.start += 2;
                    return length;
                }
            }
        }
        return undefined;
    }

    function preDecode(input, start=0) {
        return {
            data: input,
            start: start,
        };
    }

    exports.addNumb = addNumb;
    exports.addString = addString;
    exports.addFloat = addFloat;
    exports.addArray = addArray;
    exports.expectNumb = expectNumb;
    exports.expectString = expectString;
    exports.expectArray = expectArray;
    exports.preDecode = preDecode;
    exports.addPosNegNumb = addPosNegNumb;
    exports.expectPosNegNumb = expectPosNegNumb;
    exports.addHit = addHit;
    exports.expectHit = expectHit;
    exports.addPing = addPing;
    exports.expectPing = expectPing;
    exports.PLAYER_UPDATE_DELIM = PLAYER_UPDATE_DELIM;
    exports.PLAYER_DESTROY_DELIM = PLAYER_DESTROY_DELIM;

    exports.CLICK = CLICK;
    exports.ROTATE = ROTATE;
    exports.SELITEM = SELITEM;
    exports.LEFT = LEFT;
    exports.RIGHT = RIGHT;
    exports.UP = UP;
    exports.DOWN = DOWN;
    exports.UPPERLEFT = UPPERLEFT;
    exports.UPPERRIGHT = UPPERRIGHT;
    exports.LOWERLEFT = LOWERLEFT;
    exports.LOWERRIGHT = LOWERRIGHT;
    exports.MOVERELEASE = MOVERELEASE;
    exports.JOIN = JOIN;
    exports.STRUCTURE_DELIM = STRUCTURE_DELIM;
    exports.STRUCTURE_DESTROY_DELIM = STRUCTURE_DESTROY_DELIM;
    exports.PLAYER_NAME_UPDATE = PLAYER_NAME_UPDATE;
    exports.PING = PING;
    exports.CHAT = CHAT;
    exports.TOGGLEAUTOHIT = TOGGLEAUTOHIT;
    exports.BULLET_DELIM = BULLET_DELIM;
    exports.BULLET_DESTROY_DELIM = BULLET_DESTROY_DELIM;
    exports.NEW_CLAN_DELIM = NEW_CLAN_DELIM;
    exports.CLAN_DESTROY_DELIM = CLAN_DESTROY_DELIM;
    exports.CLAN_REQUEST_DELIM = CLAN_REQUEST_DELIM;
})((typeof module === 'undefined') ? window.msgpack = {} : module.exports); 